### DISCUSSION TITLE
Unable to build QPDF with emscripten

### DISCUSSION THREAD

**Discussion by bragle** [emscripten-core/emscripten#17405] [Q&A] [answered]
I have wrestled with this for almost a week now, so I guess it's time to finally admit defeat and ask for help. I hope this is the right place for that, otherwise I apologize!

I'm trying to build [QPDF](https://github.com/qpdf/qpdf) with emscripten. I found a [very old build](https://github.com/jrmuizel/qpdf.js) by @jrmuizel that works fine, but I'd like to use some of the newer functionality (you can see it in use [here](https://devina.io/pdf-editor)). I have never worked with any C language or emscripten before, so I'm sorry if there's an obvious solution to my issue.

What I've tried is the following:

1. Cloned QPDF and emscripten.
2. Copied [external libs](https://github.com/qpdf/external-libs/releases) into the QPDF root folder. This is because cmake is unable to find `zlib` and `libjpeg` when running `emcmake cmake -S . -B build -DCMAKE_BUILD_TYPE=RelWithDebInfo`. When running said command without `emcmake`, everything works fine. The documentation for using `external-libs` is in the [windows readme](https://github.com/qpdf/qpdf/blob/main/README-windows.md), so it's probably not ideal to use that on Linux, but that'll at least fix the `zlib` and `libjpeg`-issue.
3. Disabled [`BUILD_SHARED_LIBS`](https://github.com/qpdf/qpdf/blob/main/CMakeLists.txt#L70) and enabled [`USE_INSECURE_RANDOM`](https://github.com/qpdf/qpdf/blob/main/CMakeLists.txt#L73) and [`SKIP_OS_SECURE_RANDOM`](https://github.com/qpdf/qpdf/blob/main/CMakeLists.txt#L74). When running just `cmake`, leaving everything as default works just fine.
4. Changed from `if(WIN32 AND (EXISTS ${qpdf_SOURCE_DIR}/external-libs))` to `if(EXISTS ${qpdf_SOURCE_DIR}/external-libs)` in [libqpdf/CMakeLists.txt on line 129](https://github.com/qpdf/qpdf/blob/main/libqpdf/CMakeLists.txt#L129) to get the second step to actually work.
5. Create the build directory with `emcmake cmake -S . -B build -DCMAKE_BUILD_TYPE=RelWithDebInfo`.
6. Run `cmake --build build`.

Now here's the part that I'm stuck on. `wasm-ld` is not able to find external windows libraries:
```
wasm-ld: error: unable to find library -lmsvcrt
wasm-ld: error: unable to find library -lws2_32
wasm-ld: error: unable to find library -lshell32
wasm-ld: error: unable to find library -ladvapi32
wasm-ld: error: unable to find library -lgdi32
wasm-ld: error: unable to find library -luser32
wasm-ld: error: unable to find library -lcrypt32
```

I have tried building with Windows, WSL (Debian 11) and Debian 11 on a VPS. The missing libraries are all in `/mnt/c/Windows/system32` (when using WSL). What I think *might* work is maybe adding `-L/mnt/c/Windows/system32` when running the linker, but I'm not sure if that's possible. The same could potentially also be used to avoid some of the hacky things I have done to get this far such as installing `external-libs` (which AFAIK is only meant for Windows) and enabling `USE_INSECURE_RANDOM`.

This is how the linker is being executed, by the way:
```
/home/bragle/emsdk/upstream/bin/wasm-ld -o qpdf.wasm CMakeFiles/qpdf.dir/qpdf.cc.o -L/home/bragle/qpdf/external-libs/lib-mingw32 ../libqpdf/libqpdf.a /home/bragle/qpdf/external-libs/lib-mingw32/libz.a /home/bragle/qpdf/external-libs/lib-mingw32/libjpeg.a /home/bragle/qpdf/external-libs/lib-mingw32/libssl.a /home/bragle/qpdf/external-libs/lib-mingw32/libcrypto.a -lmsvcrt -lws2_32 -lshell32 -ladvapi32 -lgdi32 -luser32 -lcrypt32 -L/home/bragle/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten -lGL -lal -lhtml5 -lstubs -lnoexit -lc -ldlmalloc -lcompiler_rt -lc++-noexcept -lc++abi-noexcept -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --import-undefined --export-if-defined=main --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__main_argc_argv --export=stackSave --export=stackRestore --export=stackAlloc --export=__wasm_call_ctors --export=__errno_location --export=malloc --export=free --export=saveSetjmp --export=setThrew --export=__cxa_is_pointer_type --export-table -z stack-size=5242880 --initial-memory=16777216 --no-entry --max-memory=16777216 --global-base=1024
```

When running `cmake -S . -B build -DCMAKE_BUILD_TYPE=RelWithDebInfo` and then `cmake --build build`, it works just fine. No need for `external-libs` or changing anything in any of the `CMakeLists.txt`-files.

This is the whole error message (Windows 11, WSL2, Debian 11):
```
[ 34%] Linking CXX executable qpdf.js
wasm-ld: error: unable to find library -lmsvcrt
wasm-ld: error: unable to find library -lws2_32
wasm-ld: error: unable to find library -lshell32
wasm-ld: error: unable to find library -ladvapi32
wasm-ld: error: unable to find library -lgdi32
wasm-ld: error: unable to find library -luser32
wasm-ld: error: unable to find library -lcrypt32
em++: error: '/home/bragle/emsdk/upstream/bin/wasm-ld -o qpdf.wasm CMakeFiles/qpdf.dir/qpdf.cc.o -L/home/bragle/qpdf/external-libs/lib-mingw32 ../libqpdf/libqpdf.a /home/bragle/qpdf/external-libs/lib-mingw32/libz.a /home/bragle/qpdf/external-libs/lib-mingw32/libjpeg.a /home/bragle/qpdf/external-libs/lib-mingw32/libssl.a /home/bragle/qpdf/external-libs/lib-mingw32/libcrypto.a -lmsvcrt -lws2_32 -lshell32 -ladvapi32 -lgdi32 -luser32 -lcrypt32 -L/home/bragle/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten -lGL -lal -lhtml5 -lstubs-debug -lnoexit -lc-debug -ldlmalloc -lcompiler_rt -lc++-noexcept -lc++abi-noexcept -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --import-undefined --strip-debug --export-if-defined=main --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__wasm_call_ctors --export=__errno_location --export=malloc --export=free --export=saveSetjmp --export=setThrew --export=__cxa_is_pointer_type --export-table -z stack-size=5242880 --initial-memory=16777216 --no-entry --max-memory=16777216 --global-base=1024' failed (returned 1)
gmake[2]: *** [qpdf/CMakeFiles/qpdf.dir/build.make:101: qpdf/qpdf.js] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:438: qpdf/CMakeFiles/qpdf.dir/all] Error 2
gmake: *** [Makefile:166: all] Error 2
```

Again, I hope this is the right place for this. Otherwise, I would greatly appreciate a pointer in the right direction :smile_cat:

Many thanks!

**Comment by sbc100** [comment, marked as answer]
Emscripten cannot use any of those libraries in /mnt/c/Windows/system32 because they are win32/x86 libraries. 

You most likely need to find the place in the cmake files that is adding those libraries and somehow prevent them from being added.  Emscripten can only use emscripten-built libraries, not OS/host specific libraries such as msvct.

**Reply by bragle** [reply]
Gotcha. I realized there's probably a good reason why there's no reference to the external files anywhere else than in the windows readme, so I suspect that might be the culprit.

Thank you!
